homeassistant/components/volvo/coordinator.py

```diff
@@ -0,0 +1,216 @@
+        coordinator.async_config_entry_first_refresh.
+        """
+        _LOGGER.debug("%s - Setting up", self.config_entry.entry_id)
+
+        try:
+            vehicle = await self.api.async_get_vehicle_details()
+        except VolvoAuthException as ex:
+            raise ConfigEntryAuthFailed(
+                translation_domain=DOMAIN,
+                translation_key="unauthorized",
+                translation_placeholders={"message": ex.message},
+            ) from ex
+
+        if vehicle is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN, translation_key="no_vehicle"
+            )
+
+        self.vehicle = vehicle
+        self.data = {}
+
+        device_name = (
+            f"{MANUFACTURER} {vehicle.description.model} {vehicle.model_year}"
+            if vehicle.fuel_type == "NONE"
+            else f"{MANUFACTURER} {vehicle.description.model} {vehicle.fuel_type} {vehicle.model_year}"
+        )
+
+        self.device = DeviceInfo(
+            identifiers={(DOMAIN, vehicle.vin)},
+            manufacturer=MANUFACTURER,
+            model=f"{vehicle.description.model} ({vehicle.model_year})",
+            name=device_name,
+            serial_number=vehicle.vin,
+        )
+
+        self.hass.config_entries.async_update_entry(
+            self.config_entry,
+            title=f"{MANUFACTURER} {vehicle.description.model} ({vehicle.vin})",
+        )
+
+        self._refresh_conditions = {
+            "command_accessibility": (self.api.async_get_command_accessibility, True),
+            "diagnostics": (self.api.async_get_diagnostics, True),
+            "fuel": (
+                self.api.async_get_fuel_status,
+                self.vehicle.has_combustion_engine(),
+            ),
+            "odometer": (self.api.async_get_odometer, True),
+            "recharge_status": (
+                self.api.async_get_recharge_status,
+                self.vehicle.has_battery_engine(),
+            ),
+            "statistics": (self.api.async_get_statistics, True),
+        }
+
+    async def _async_update_data(self) -> CoordinatorData:
+        """Fetch data from API."""
+        _LOGGER.debug("%s - Updating data", self.config_entry.entry_id)
+
+        api_calls = self._get_api_calls()
+        data: CoordinatorData = {}
+        valid = 0
+        exception: Exception | None = None
+
+        results = await asyncio.gather(
+            *(call() for call in api_calls), return_exceptions=True
+        )
+
+        for result in results:
+            if isinstance(result, VolvoAuthException):
+                # If one result is a VolvoAuthException, then probably all requests
+                # will fail. In this case we can cancel everything to
+                # reauthenticate.
+                #
+                # Raising ConfigEntryAuthFailed will cancel future updates
+                # and start a config flow with SOURCE_REAUTH (async_step_reauth)
```

Comment:
Given the below comments I wonder if this is the right approach?
It seems any random call can be 404 while the others not so isn't then raising a reauth flow the wrong approach (as there isn't really a problem with the key)?

Should not all calls be returned with `VolvoAuthException` before raising?

Reply:
Right, I got confused about the comment below on 404.